casper/src/main/resources/PoS.rhox

```diff
@@ -63,626 +66,646 @@ in {
+                                      accumulateDeposits!(*accDepositsDoneCh) |
+                                      for (_ <- accDepositsDoneCh) {
+                                        new pendingRewardsAllCh in {
+                                          // Collects pending rewards from each validator and resets to zero.
+                                          PendingRewards!("collect", state.get("activeValidators"), *pendingRewardsAllCh) |
+                                          for (@pendingRewards <- pendingRewardsAllCh) {
+                                            new ackInitCh in {
+                                              // Initialize empty pending rewards state with current set of validators.
+                                              PendingRewards!("init", state.get("activeValidators"), *ackInitCh) |
+                                              for (_ <- ackInitCh) {
+                                                // Add all pending rewards to global committed rewards.
+                                                commitRewards!(pendingRewards, state.get("committedRewards"), *commitRewardsCh)
+                                              }
+                                            }
+                                          }
+                                        } |
+                                        for (@committedRewardsList <- commitRewardsCh) {
+                                          // Collects slashed validator's committed rewards and bond.
+                                          valRewardsCh!(committedRewardsList.toMap().get(validator)) |
+                                          valBondCh!(state.get("allBonds").get(validator)) |
+                                          for (@valRewards <- valRewardsCh;
+                                               @valBond    <- valBondCh) {
+                                            // Transfers slashed funds from PoS vault to Coop vault, then updates state.
+                                            new transferDoneCh in {
+                                              @posVault!(
+                                                "transfer",
+                                                coopRevAddress,
+                                                valRewards + valBond,
+                                                posAuthKey,
+                                                *transferDoneCh
+                                              ) |
+                                              for (_ <- transferDoneCh) {
+                                                // Sets slashed validator's committed rewards and bond to 0 in state.
+                                                slashResultCh!({
+                                                  "allBonds"        : state.get("allBonds").set(validator, 0),
+                                                  "activeValidators": state.get("activeValidators"),
```

Comment:
When is slashed validator removed from the list of active validators?

Reply:
This is something that still needs to be decided. In this change, the slashed validator's bond is transferred to the Coop vault and set to 0 so they will not receive any rewards because those are stake-weighted. I think the right answer is either to immediately move them to the withdrawers map (not sure how this would be rectified if the slash was found to be a mistake) or to include a slashed map in the global state with some grace period before they are moved to withdrawers.

Reply:
Slashed validator should be immediately removed from active validators list if fully slashed

Reply:
Slashed validator is immediately placed in the withdrawers map.